src/test/java/de/tum/in/www1/artemis/service/notifications/GroupNotificationServiceTest.java

```diff
@@ -117,95 +96,61 @@
     private final static int NUMBER_OF_ALL_GROUPS = 4;
 
     /**
-     * Sets up all needed mocks and their wanted behavior once for all test cases.
-     * These are the common mocks and structures which behavior is fixed and will not change
+     * Sets up all needed mocks and their wanted behavior.
      */
-    @BeforeAll
-    public static void setUp() {
-        mailService = mock(MailService.class);
-        doNothing().when(mailService).sendNotificationEmailForMultipleUsers(any(), any(), any());
-
-        course = new Course();
-        course.setId(COURSE_ID);
-
-        user = new User();
-
-        users.add(user);
-
-        userRepository = mock(UserRepository.class);
-        when(userRepository.getStudents(any())).thenReturn(users);
-        when(userRepository.getInstructors(any())).thenReturn(users);
-        when(userRepository.getEditors(any())).thenReturn(users);
-
-        notificationCaptor = ArgumentCaptor.forClass(Notification.class);
-
-        groupNotificationRepository = mock(GroupNotificationRepository.class);
-        when(groupNotificationRepository.save(any())).thenReturn(null);
+    @BeforeEach
+    public void setUp() {
+        course = database.createCourse();
 
-        messagingTemplate = mock(SimpMessageSendingOperations.class);
+        database.addUsers(1, 0, 0, 1);
 
-        notificationSettingsService = mock(NotificationSettingsService.class);
+        student = database.getUserByLogin("student1");
 
-        groupNotificationService = spy(new GroupNotificationService(groupNotificationRepository, messagingTemplate, userRepository, mailService, notificationSettingsService));
+        instructor = database.getUserByLogin("instructor1");
 
         archiveErrors = new ArrayList<>();
 
-        course = new Course();
-        course.setId(COURSE_ID);
-
-        exam = new Exam();
-        exam.setId(EXAM_ID);
-        exam.setCourse(course);
+        exam = database.addExam(course);
+        examRepository.save(exam);
 
         lecture = new Lecture();
         lecture.setCourse(course);
 
         attachment = new Attachment();
 
-        exercise = new TextExercise();
-        updatedExercise = new TextExercise();
+        exercise = ModelFactory.generateTextExercise(null, null, null, course);
+        exerciseRepository.save(exercise);
+        updatedExercise = ModelFactory.generateTextExercise(null, null, null, course);
+        exerciseRepository.save(updatedExercise);
 
-        exerciseGroup = new ExerciseGroup();
+        ExerciseGroup exerciseGroup = new ExerciseGroup();
         exerciseGroup.setExam(exam);
 
         examExercise = new TextExercise();
         examExercise.setExerciseGroup(exerciseGroup);
         examExercise.setProblemStatement(EXAM_PROBLEM_STATEMENT);
 
-        quizExercise = new QuizExercise();
-        quizExercise.setCourse(course);
+        quizExercise = database.createQuiz(course, null, null);
```

Comment:
I'm adding the comment here, but I'm actually referring to line 129-131 in your changes. Is there a particular reason you decided not to use `Modelfactory.generateTextExerciseForExam()`?

Reply:
Sure, good question!
From my understanding (and the fact that we have no real convention on how to do this (no information about this in our documentation)) this is basically the same thing as long as you do not mock it.
We prefer to use real entity-type objects and how you create them is not as important.
E.g. in this case the NotificationServices use e.g. exercises as a transient object (it is not pulled or saved/changed to the DB), i.e. it works just as any other input. Therefore, It is enough to have the needed properties set and not a lot more than the database function provide
![image](https://user-images.githubusercontent.com/65814168/147880381-1f426bdf-abd8-4b19-9ef0-8603761c4ef3.png)
![image](https://user-images.githubusercontent.com/65814168/147880382-677c10cb-d72f-4779-8b02-5717146703a2.png)

Reply:
I understand. As I already said, in your case the use of the `Modelfactory` would not simplify the code critically imo. Therefore your way is fine for me.